docs/namespace_config_spec.md

```diff
@@ -0,0 +1,309 @@
+
+The following declares a type *Document* with three relations: *owners* and
+*viewers*, both of which have *users* as subjects. Additionally, the relation
+*parent* has type *Document*.
+```ts
+type User { metadata = { id: 1 } }
+
+type Document {
+  metadata = { id: 2 }
+  related = {
+    parent: Document[]
+    owners: User[]
+    viewers: User[]
+  }
+}
+```
+
+### Permission definition
+
+Permissions are defined as functions within class declarations 
+that take a parameter `ctx` of type `Context` and evaluate to a boolean `true` or
+`false`.
+
+```ebnf
+PermissionDefns = "permits" "=" "{" Permission [ "," Permission ] "}" .
+Permission      = PermissionSign "=>" PermissionBody .
+PermissionSign  = PermissionName ":" "(" "ctx" [ ":" "Context" ] ")" [ ":" "boolean" ] .
+PermissionName  = identifier .
+```
+
+The `ctx` object is a fixed parameter that contains the `subject` for which the
+permission check should be conducted:
+```typescript
+ctx = {subject: "some_user_id"}
+```
+In the future, more context fields could be added here, e.g., the IP range or
+geolocation, the time of day, or the security level of the device making the
+request.
+
+
+```ebnf
+PermissionBody  = PermissionCheck | { Operator PermissionCheck } .
+Operator        = "||" | "&&" .
+PermissionCheck = TransitiveCheck | IncludesCheck .
+```
+
+The body of a permission check is either one of
+
+* a `IncludesCheck`, a check that something is in a set, e.g., `this.related.viewers.includes(ctx.subject)`:
+```ebnf
+IncludesCheck   = "this" "." "related" "." RelationName "." "includes" "(" "ctx" "." "subject" ")" .
+```
+* a `TranstitiveCheck`, a call to a permission on a relation, e.g., `this.related.parents.some(p => p.permits.view(ctx))`:
+```ebnf
+TransitiveCheck = "this" "." "related" "." RelationName "." "some" "(" Var "=>" PermissionCall ")" .
+PermissionCall  = Var "." "permits" "." PermissionName "(" "ctx" ")" .
+Var             = identifier .
+```
+
+#### Implementation note
+`IncludeCheck` and `TransitiveCheck` will be translated to Zanzibar concepts as
+follows:
+
+| Keto Config | Zanzibar AST |
+| --- | --- |
+| `this.related.<R>.includes(ctx.subject)` | `computed_userset { relation: "<R>" } }` |
+| `this.related.<R>.some(x => x.permits.<P>(ctx.subject))` | `tuple_to_userset { tupleset { relation: "<R>" } computed_userset { relation: "<P>" } }` |
+
+## Examples
+
+The config can be typechecked in `strict` mode by TypeScript with the `noLib`
+option (preventing the standard globals), the `strictPropertyInitialization`
+option (allowing uninitialized properties) and supplying the following globals
+instead:
+
+```typescript
+type Context = { subject: never }
+
+interface Namespace {
+    metadata: { id: number }
+    related?: { [relation: string]: Namespace[] }
+    permits?: { [method: string]: (ctx: Context) => boolean }
+}
+
+interface Array<Namespace> {
+  includes(element: Namespace): boolean;
+  some(callbackfn: (value: Namespace) => boolean): boolean;
+}
```

Comment:
This part is already listed in identical form on lines 105-116. Can we remove one of them?

Reply:
yes